hibernate-infinispan/src/main/java/org/hibernate/cache/infinispan/query/QueryResultsRegionImpl.java

```diff
@@ -1,79 +1,209 @@
       boolean skipCacheStore = false;
-      if (!isValid())
+      if ( !isValid() ) {
          skipCacheStore = true;
+      }
 
-      if (!checkValid())
+      if ( !checkValid() ) {
          return null;
+      }
 
       // In Infinispan get doesn't acquire any locks, so no need to suspend the tx.
       // In the past, when get operations acquired locks, suspending the tx was a way
       // to avoid holding locks that would prevent updates.
       // Add a zero (or low) timeout option so we don't block
       // waiting for tx's that did a put to commit
-      if (skipCacheStore)
-         return getCache.withFlags(Flag.SKIP_CACHE_STORE).get(key);
-      else
-         return getCache.get(key);
-   }   
+      TransactionManager tm = getTransactionManager();
+      try {
+         if (tm != null && tm.getStatus() == Status.STATUS_ACTIVE) {
+            Transaction transaction = tm.getTransaction();
+            if (transaction != null) {
+               Map<Object, PostTransactionQueryUpdate> map = transactionContext.get(transaction);
+               if (map != null) {
+                  PostTransactionQueryUpdate update = map.get(key);                  
+                  if (update != null) {
+                     Object value = update.getValue();
+                     if (value != null) {
+                        return value;
+                     }
+                  }
+               }
+            }
+         }
+      } catch (SystemException e) {
+         log.trace("Failed to retrieve current transaction status.", e);
+      }
+      if ( skipCacheStore ) {
+         return getCache.withFlags( Flag.SKIP_CACHE_STORE ).get( key );
+      }
+      else {
+         return getCache.get( key );
+      }      
+   }
 
+   @Override
+   @SuppressWarnings("unchecked")
    public void put(Object key, Object value) throws CacheException {
-      if (checkValid()) {
+      if ( checkValid() ) {
+         // See HHH-7898: Even with FAIL_SILENTLY flag, failure to write in transaction
+         // fails the whole transaction. It is an Infinispan quirk that cannot be fixed
+         // ISPN-5356 tracks that. This is because if the transaction continued the
+         // value could be committed on backup owners, including the failed operation,
+         // and the result would not be consistent.
+         TransactionManager tm = getTransactionManager();
+         Transaction transaction = null;
+         try {
+            transaction = tm != null && tm.getStatus() == Status.STATUS_ACTIVE ? tm.getTransaction() : null;
+            if (transaction != null) {
+               Map<Object, PostTransactionQueryUpdate> map = transactionContext.get(transaction);
+               if (map == null) {
+                  // The map implementation needs to be synchronized as we can access that from evict, too.
+                  transactionContext.put(transaction, map = new ConcurrentHashMap());
```

Comment:
Again, flaw in master discovered during the review process. `evict()` should update the values in `transactionContext`, too, and therefore there can be concurrent access to it. As the comment one line above says.